ipfs-4-to-5/migration/migration.go

```diff
@@ -0,0 +1,211 @@
+package mg3
+
+import (
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"strconv"
+
+	migrate "github.com/ipfs/fs-repo-migrations/go-migrate"
+	lock "github.com/ipfs/fs-repo-migrations/ipfs-1-to-2/repolock"
+	mfsr "github.com/ipfs/fs-repo-migrations/mfsr"
+	log "github.com/ipfs/fs-repo-migrations/stump"
+
+	flatfs "github.com/ipfs/go-ds-flatfs"
+)
+
+type Migration struct{}
+
+func (m Migration) Versions() string {
+	return "4-to-5"
+}
+
+func (m Migration) Reversible() bool {
+	return true
+}
+
+func revertStep2(ffspath string) error {
+	if err := flatfs.DowngradeV1toV0(ffspath); err != nil {
+		return fmt.Errorf("reverting flatfsv1 upgrade: %s", err)
+	}
+	return nil
+}
+
+func (m Migration) Apply(opts migrate.Options) error {
+	log.Verbose = opts.Verbose
+	log.Log("applying %s repo migration", m.Versions())
+
+	log.VLog("locking repo at %q", opts.Path)
+	lk, err := lock.Lock2(opts.Path)
+	if err != nil {
+		return err
+	}
+	defer lk.Close()
+
+	repo := mfsr.RepoPath(opts.Path)
+
+	log.VLog("  - verifying version is '3'")
+	if err := repo.CheckVersion("4"); err != nil {
+		return err
+	}
+
+	basepath := filepath.Join(opts.Path, "blocks")
+	ffspath := filepath.Join(opts.Path, "blocks-v4")
+	if err := os.Rename(basepath, ffspath); err != nil {
+		return err
+	}
+
+	log.Log("Upgrading datastore format to have sharding specification file")
+	if err := flatfs.UpgradeV0toV1(ffspath, 5); err != nil {
+		return err
+	}
+
+	tempffs := filepath.Join(opts.Path, "blocks-v5")
+	log.Log("creating a new flatfs datastore with new format")
+	if err := flatfs.Create(tempffs, flatfs.NextToLast(2)); err != nil {
+		if err2 := revertStep2(ffspath); err2 != nil {
+			log.Error(err2)
+		}
+		return err
+	}
+
+	revert3 := func(mainerr error) error {
+		log.Error("failed to convert flatfs datastore: %s", mainerr)
+		log.Log("attempting to revert...")
```

Comment:
This might not be the best option, espacally if the process fails halfway though and can easily be corrected.  It way be better to give to allow the conversion to be restartable (you would have to skip the initial steps) and just leave things as is so the user can fix the problem and try again.

That being said.  I'm okay with leaving things as is in the interest of getting something out quickly.